x-pack/filebeat/input/httpjson/input.go

```diff
@@ -81,532 +48,145 @@ func newRetryLogger() *retryLogger {
 
-	var dateStr string
-	if in.nextCursorValue == "" {
-		t := timeNow().UTC().Add(-in.config.DateCursor.InitialInterval)
-		dateStr = t.Format(in.config.DateCursor.GetDateFormat())
-	} else {
-		dateStr = in.nextCursorValue
-	}
+	stdCtx := ctxtool.FromCanceller(ctx.Cancelation)
 
-	url, err := url.Parse(in.config.URL)
+	httpClient, err := in.newHTTPClient(stdCtx)
 	if err != nil {
-		return in.config.URL
+		return err
 	}
 
-	q := url.Query()
-
-	var value string
-	if in.config.DateCursor.ValueTemplate == nil {
-		value = dateStr
-	} else {
-		buf := new(bytes.Buffer)
-		if err := in.config.DateCursor.ValueTemplate.Execute(buf, dateStr); err != nil {
-			return in.config.URL
-		}
-		value = buf.String()
-	}
+	dateCursor := newDateCursorFromConfig(in.config, log)
 
-	q.Set(in.config.DateCursor.URLField, value)
+	rateLimiter := newRateLimiterFromConfig(in.config, log)
 
-	url.RawQuery = q.Encode()
+	pagination := newPaginationFromConfig(in.config)
 
-	return url.String()
-}
+	requester := newRequester(
+		in.config,
+		rateLimiter,
+		dateCursor,
+		pagination,
+		httpClient,
+		log,
+	)
 
-func (in *HttpjsonInput) advanceCursor(m common.MapStr) {
-	if in.config.DateCursor.Field == "" {
-		in.nextCursorValue = time.Now().UTC().Format(in.config.DateCursor.GetDateFormat())
-		return
+	if err := requester.processHTTPRequest(stdCtx, publisher); err != nil {
+		return err
 	}
 
-	v, err := m.GetValue(in.config.DateCursor.Field)
-	if err != nil {
-		in.log.Warnf("date_cursor field: %q", err)
-		return
-	}
-	switch t := v.(type) {
-	case string:
-		_, err := time.Parse(in.config.DateCursor.GetDateFormat(), t)
-		if err != nil {
-			in.log.Warn("date_cursor field does not have the expected layout")
-			return
-		}
-		in.nextCursorValue = t
-	default:
-		in.log.Warn("date_cursor field must be a string, cursor will not advance")
-		return
+	if in.config.Interval == 0 {
+		return nil
 	}
-}
 
-func (in *HttpjsonInput) run() error {
-	ctx, cancel := context.WithCancel(in.workerCtx)
-	defer cancel()
+	err = timed.Periodic(stdCtx, in.config.Interval, func() error {
+		log.Info("Process another repeated request.")
+		return requester.processHTTPRequest(stdCtx, publisher)
+	})
 
-	client, err := in.newHTTPClient(ctx)
-	if err != nil {
-		return err
+	if err == nil {
+		return nil
 	}
 
-	ri := &RequestInfo{
-		ContentMap: common.MapStr{},
-		Headers:    in.HTTPHeaders,
-	}
-	if in.config.HTTPMethod == "POST" && in.config.HTTPRequestBody != nil {
-		ri.ContentMap.Update(common.MapStr(in.config.HTTPRequestBody))
-	}
-	err = in.processHTTPRequest(ctx, client, ri)
-	if err == nil && in.Interval > 0 {
-		ticker := time.NewTicker(in.Interval)
-		defer ticker.Stop()
-		for {
-			select {
-			case <-ctx.Done():
-				in.log.Info("Context done.")
-				return nil
-			case <-ticker.C:
-				in.log.Info("Process another repeated request.")
-				err = in.processHTTPRequest(ctx, client, ri)
-				if err != nil {
-					return err
-				}
-			}
-		}
+	switch errors.Cause(err) {
```

Comment:
Cause is used by some legacy error libs (e.g. pkg/errors which is not maintained anymore). Better use `Unpack` or `errors.Is(context.X)`.